src/object/srv_obj.c

```diff
@@ -943,126 +1029,43 @@ ds_obj_rw_handler(crt_rpc_t *rpc)
 				DP_UOID(orw->orw_oid), rc);
 			D_GOTO(out, rc);
 		}
-
-		dti_cos = dth->dth_dti_cos;
-		dti_cos_count = dth->dth_dti_cos_count;
-		if (!(orw->orw_flags & ORF_FROM_LEADER) && !dispatch)
-			dth->dth_non_rep = 1;
 	}
-
 	D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_PREP);
 
-dispatch:
-	if (dispatch) {
-		D_TIME_START(tls->ot_sp, OBJ_PF_UPDATE_DISPATCH);
-
-		rc = ds_obj_req_disp_prepare(rpc->cr_opc,
-				orw->orw_shard_tgts.ca_arrays,
-				orw->orw_shard_tgts.ca_count,
-				obj_update_prefw, obj_update_postfw, orw, flags,
-				dti_cos_count, dti_cos, &obj_arg);
-		if (rc != 0) {
-			D_ERROR(DF_UOID": ds_obj_req_disp_prepare failed %d.\n",
-				DP_UOID(orw->orw_oid), rc);
-			D_GOTO(out, rc);
-		}
-		D_ASSERT(obj_arg != NULL);
-
-		rc = dss_ult_create(ds_obj_req_dispatch, obj_arg,
-				    DSS_ULT_IOFW, tag, 0, NULL);
-		if (rc != 0) {
-			D_ERROR(DF_UOID": ds_obj_req_dispatch failed %d.\n",
-				DP_UOID(orw->orw_oid), rc);
-			ds_obj_req_disp_arg_free(obj_arg);
-			D_GOTO(out, rc);
-		}
-
-		D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_DISPATCH);
-	}
-
-	if (!resend) {
-		D_TIME_START(tls->ot_sp, OBJ_PF_UPDATE_LOCAL);
-
-		/* local RPC handler */
-		rc = ds_obj_rw_local_hdlr(rpc, tag, cont_hdl, cont,
-					  &ioh, true, map_ver, dth);
-		if (rc != 0)
-			D_ERROR(DF_UOID": rw_local_hdlr (update) failed %d.\n",
-				DP_UOID(orw->orw_oid), rc);
-
-		D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_LOCAL);
-	}
-
-	/* wait dispatched IO's completion when needed */
-	if (dispatch) {
-		struct dtx_conflict_entry	*dces = NULL;
-
-		D_TIME_START(tls->ot_sp, OBJ_PF_UPDATE_WAIT);
-
-		rc1 = ds_obj_req_disp_wait(obj_arg, rc >= 0 ? &dces : NULL);
-		if (dces != NULL) {
-			/* XXX: The local modification has been done, but remote
-			 *	replica failed because of some uncommitted DTX,
-			 *	it may be caused by some garbage DTXs on remote
-			 *	replicas or leader has more information because
-			 *	of CoS cache. So handle (abort or commit) them
-			 *	firstly then retry.
-			 */
-
-			D_ASSERT(dth != NULL);
-
-			D_DEBUG(DB_TRACE, "Hit conflict DTX (%d)"DF_DTI" for "
-				DF_DTI", handle them and retry update.\n",
-				rc1, DP_DTI(&dces[0].dce_xid),
-				DP_DTI(&orw->orw_dti));
-
-			rc = dtx_conflict(cont->sc_hdl, dth,
-					  cont_hdl->sch_pool->spc_uuid,
-					  cont->sc_uuid, dces, rc1,
-					  cont_hdl->sch_pool->spc_map_version);
-			D_FREE(dces);
-			if (rc >= 0) {
-				resend = true;
-				flags |= ORF_RESEND;
-				dti_cos = dth->dth_dti_cos;
-				dti_cos_count = dth->dth_dti_cos_count;
-
-				D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_WAIT);
-
-				goto dispatch;
-			}
-		} else if (rc >= 0) {
-			rc = rc1;
-		}
-
-		D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_WAIT);
-	}
-
+	exec_arg.rpc = rpc;
+	exec_arg.cont_hdl = cont_hdl;
+	exec_arg.cont = cont;
+again:
+	exec_arg.flags = flags;
+	/* Execute the operation on all targets */
+	rc = dtx_exec_ops(orw->orw_shard_tgts.ca_arrays,
+			  orw->orw_shard_tgts.ca_count, dth,
+			  ds_obj_tgt_update, &exec_arg);
```

Comment:
here seems some error case need to be checked/refined between dtx_exec_ops()'s failure and dtx_end()'s return code. I just found two cases:
1) if dtx_exec_ops just failed as ENOMEM when allocate dtx_exec_arg, then dtx_end() -> dtx_wait() just returns zero but following handing in dtx_end() seems incorrect.
2) if the future create failed in dtx_exec_ops(), then it will cause assertion in dtx_wait().

Reply:
ah, right. good catch